flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java

```diff
@@ -51,14 +52,15 @@
  *       <li>Store the meta of files into {@link ChangelogTaskLocalStateStore} by
  *           AsyncCheckpointRunnable#reportCompletedSnapshotStates().
  *       <li>Pass control of the file to {@link LocalChangelogRegistry#register} when
- *           ChangelogKeyedStateBackend#notifyCheckpointComplete() , files of the previous
- *           checkpoint will be deleted by {@link LocalChangelogRegistry#discardUpToCheckpoint} at
- *           the same time.
+ *           FsStateChangelogWriter#persist , files of the previous checkpoint will be deleted by
+ *           {@link LocalChangelogRegistry#discardUpToCheckpoint} when the previous checkpoint is
+ *           confirmed.
```

Comment:
Yes, the cleaning up of local files depends on the newly completed checkpoint, local files will accumulate if no checkpoint could complete. IIUC, remote dstl file has same problem. I did encounter this problem in our environment. 
<img width="683" alt="image" src="https://user-images.githubusercontent.com/16664055/232822869-ae8ccf9f-3caf-42d2-9d5b-1d19523612cd.png">
I suggest open a new ticket to solve this problem both in remote and local, WDYT @Myasuka ?

Reply:
I think the problem does not exist for local checkpoints before this PR. If so, I am against to introduce anther new problem here.

Reply:
The problem of not cleaning up useless local files after truncate was not actively addressed before this PR. Previously, it was only aggressive to delete all files referenced by checkpoints that aborted.

Reply:
The current implementation can partially alleviate the issue of local file accumulation:
1. In the case that TM can exit normally(`ChangelogTaskLocalStateStore#dispose()` can be executed), all local files can be cleaned up. 
2. If `emptyDir` is used as the local disk, all files will be deleted when the pod exits.

So I think the problem with local dstl files is milder than remote dstl files. 

[Pruning ](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskStateManagerImpl.java#L210) other dstl local files during restore may alleviate this issue, WDYT?

Reply:
@zoltar9264 My concern is that we already know that this solution is not a good choice, we might need to refactor this in the near feature. Why we have to do the work twice?

On the other hand, since local disk has a much smaller space than DFS. Leaving files on the disk could make no space left problem more easily, which actually make the status worse.

Reply:
Thanks for explanation @Myasuka  , I agree that one shouldn't introduce a solution that will be refactored anytime soon. 
But I just think the approach in that PR is one that will still be appropriate in the long run. We can divide the deletion of dstl files into two categories :

1. delete by subsumed checkpoint 
        remote files : delete by jobmanager , nothing with taskmanager
        local files : delete by _LocalChangelogRegistry#discardUpToCheckpoint_ (this PR does exactly that)
2. delete by not used StateChangeSet ( also contains thats only used by aborted checkpoints ) 
        remote files and local files are in the same processing logic, both are processed by _TaskChangelogRegistry_ . Perhaps we can introduce a "StateChangeUsageTracker" component to find out those not used StateChangeSets based on checkpoint completion and materialization execution, and then notify TaskChangelogRegistry to release the reference count of the corresponding file.

In the method described above, the LocalChangelogRegistry is only responsible for the deletion of local files at subsumed checkpoint, as implemented in this PR. So I suggest approve this pr and open a new ticket to address deletion of not used both remote and local files .

Reply:
I think a better way is that we should move to the Flink ticket and have a clean discussion on this problem with consideration of the whole plan, instead of just approving a half-baked PR.
What's more, we don't have a release deadline recently, and shall not be hurry and in this case.